File: modules/crispy-native-core/android/src/main/java/aayush/crispy/core/CrispyNativeCoreModule.kt
package aayush.crispy.core

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Build
import android.os.IBinder
import android.util.Log
import android.os.Looper
import android.os.Handler
import com.facebook.react.bridge.ReactContext
import aayush.crispy.core.pip.PipController
import aayush.crispy.core.player.PlayerActivity
import java.util.HashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class CrispyNativeCoreModule : Module() {
  // CrispyServer is owned exclusively by TorrentService and started on-demand.
  // This avoids EADDRINUSE when both Module and Service try to bind port 11470.
  private var torrentService: TorrentService? = null
  private var isBound = false
  private var serviceLatch = CountDownLatch(1)
  private var startupCleanupLatch = CountDownLatch(0)

  private val connection = object : ServiceConnection {
    override fun onServiceConnected(className: android.content.ComponentName, service: IBinder) {
      val binder = service as TorrentService.TorrentBinder
      torrentService = binder.getService()
      isBound = true
      serviceLatch.countDown()
      Log.d("CrispyModule", "TorrentService connected")
    }

    override fun onServiceDisconnected(arg0: android.content.ComponentName) {
      isBound = false
      torrentService = null
      serviceLatch = CountDownLatch(1)
    }
  }

  override fun definition() = ModuleDefinition {
    Name("CrispyNativeCore")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate
      startupCleanupLatch = CountDownLatch(1)

      val reactContext = context as? ReactContext
      if (reactContext != null) {
        // Centralized PiP controller (events + params updates).
        val app = reactContext.applicationContext as? android.app.Application
        if (app != null) PipController.start(app, reactContext)
      }

      Thread {
        try {
          TorrentService.cleanupTorrentStorage(context.applicationContext, "module_create")
        } catch (e: Exception) {
          Log.e("CrispyModule", "Startup torrent cleanup failed", e)
        } finally {
          startupCleanupLatch.countDown()
        }
      }.start()

      // TorrentService (with its CrispyServer) is started on-demand in ensureService()
    }

    OnDestroy {
      val context = appContext.reactContext ?: return@OnDestroy
      
      // Stop everything aggressively
      if (isBound) {
        try {
          torrentService?.stopAll(clearStorage = true)
          context.unbindService(connection)
        } catch (e: Exception) {
          Log.e("CrispyModule", "Error during OnDestroy unbind", e)
        }
        isBound = false
      }
      
      PipController.stop()
    }

    AsyncFunction("startStream") { infoHash: String, fileIdx: Int, sessionId: String ->
      Log.d("CrispyModule", "[JS] startStream: $infoHash, index: $fileIdx, session: $sessionId")

      waitForStartupCleanup()
      
      val service = ensureService()
      if (service == null) {
          Log.e("CrispyModule", "Failed to start stream: Service not bound (timeout)")
          return@AsyncFunction null
      }

      if (!service.awaitServerReady()) {
          Log.e("CrispyModule", "Failed to start stream: local server not ready")
          return@AsyncFunction null
      }
      
      // Start torrent download (non-blocking) with Session ID
      if (!service.startInfoHash(infoHash, sessionId)) {
          Log.e("CrispyModule", "Failed to start torrent: $infoHash")
          return@AsyncFunction null
      }
      
      // Return URL immediately - Native Server handles waiting for metadata if player connects early
      val idx = if (fileIdx >= 0) fileIdx else service.getLargestFileIndex(infoHash)
      val url = service.getStreamUrl(infoHash, idx)
      Log.d("CrispyModule", "[JS] -> resolved URL immediately: $url")
      return@AsyncFunction url
    }

    AsyncFunction("destroyStream") { sessionId: String ->
      Log.d("CrispyModule", "[JS] destroyStream: $sessionId")
      val service = torrentService ?: ensureService()
      if (service != null) {
          service.stopAll(sessionId, clearStorage = true)
      } else {
          val context = appContext.reactContext
          if (context != null) {
              TorrentService.cleanupTorrentStorage(context.applicationContext, "destroy_stream_unbound")
          }
      }
      
      // PRODUCTION: If no more active torrents, unbind to allow service to stop
      if (torrentService?.hasActiveTorrents() != true) {
          Log.d("CrispyModule", "No active torrents, unbinding service...")
          if (isBound) {
              val context = appContext.reactContext
              context?.unbindService(connection)
              isBound = false
              torrentService = null
              serviceLatch = CountDownLatch(1)
          }
      }
    }

    AsyncFunction("stopTorrent") { infoHash: String ->
      torrentService?.stopTorrent(infoHash)
    }

    AsyncFunction("destroyTorrent") { infoHash: String ->
      torrentService?.deleteTorrentData(infoHash)
    }

    AsyncFunction("clearCache") {
      val context = appContext.reactContext ?: return@AsyncFunction
      val service = torrentService
      if (service != null) {
        service.stopAll(clearStorage = true)
      } else {
        TorrentService.cleanupTorrentStorage(context.applicationContext, "clear_cache")
      }
    }

    AsyncFunction("handleSeek") { infoHash: String, fileIdx: Int, position: Long ->
      Log.d("CrispyModule", "[JS] handleSeek: $infoHash, index: $fileIdx, pos: $position")
      torrentService?.handleSeek(infoHash, fileIdx, position)
    }

    AsyncFunction("enterPiP") { width: Double?, height: Double? ->
      val activity = appContext.currentActivity ?: return@AsyncFunction false
      return@AsyncFunction PipController.enterPiP(activity, width, height)
    }

    /**
     * Updates PiP configuration without necessarily entering PiP.
     *
     * JS should call this from the player screen whenever:
     * - playback starts/pauses
     * - video dimensions become known
     * - player screen mounts/unmounts (enabled flag)
     */
    AsyncFunction("setPiPConfig") { enabled: Boolean, isPlaying: Boolean, width: Double?, height: Double? ->
      PipController.setConfigFromJs(enabled, isPlaying, width, height)
      return@AsyncFunction true
    }

    AsyncFunction("isInPiPMode") {
      val activity = appContext.currentActivity
      if (activity == null) return@AsyncFunction false
      return@AsyncFunction PipController.isInPiPMode(activity)
    }

    // --- NATIVE PLAYER ACTIVITY (Android) ---
    AsyncFunction("openPlayerActivity") { sessionId: String, url: String, headers: Map<String, String>?, engine: String?, paused: Boolean, title: String?, artist: String?, artworkUrl: String? ->
      val ctx = appContext.reactContext ?: return@AsyncFunction false
      val activity = appContext.currentActivity

      val intent = Intent(activity ?: ctx, PlayerActivity::class.java)
      if (activity == null) intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

      intent.putExtra(PlayerActivity.EXTRA_SESSION_ID, sessionId)
      intent.putExtra(PlayerActivity.EXTRA_URL, url)
      intent.putExtra(PlayerActivity.EXTRA_ENGINE, engine ?: PlayerActivity.ENGINE_EXO)
      intent.putExtra(PlayerActivity.EXTRA_PAUSED, paused)
      intent.putExtra(PlayerActivity.EXTRA_TITLE, title ?: "")
      intent.putExtra(PlayerActivity.EXTRA_ARTIST, artist ?: "")
      intent.putExtra(PlayerActivity.EXTRA_ARTWORK_URL, artworkUrl)
      if (headers != null) {
        intent.putExtra(PlayerActivity.EXTRA_HEADERS, HashMap(headers))
      }

      return@AsyncFunction try {
        val starter = (activity ?: ctx)
        if (Looper.myLooper() == Looper.getMainLooper()) {
          starter.startActivity(intent)
        } else {
          Handler(Looper.getMainLooper()).post {
            try {
              starter.startActivity(intent)
            } catch (t: Throwable) {
              Log.e("CrispyModule", "openPlayerActivity failed", t)
            }
          }
        }
        true
      } catch (t: Throwable) {
        Log.e("CrispyModule", "openPlayerActivity failed", t)
        false
      }
    }

    AsyncFunction("closePlayerActivity") {
      Log.i("CrispyModule", "closePlayerActivity")
      return@AsyncFunction withPlayerActivityUi("closePlayerActivity") { it.stopPlaybackAndFinishFromJs("closePlayerActivity") }
    }

    AsyncFunction("nativePlayerSetPaused") { paused: Boolean ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetPaused") { it.setPausedFromJs(paused) }
    }

    AsyncFunction("nativePlayerSeek") { positionSec: Double ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSeek") { it.seekFromJs(positionSec) }
    }

    AsyncFunction("nativePlayerLoad") { url: String?, headers: Map<String, String>?, paused: Boolean, title: String?, artist: String?, artworkUrl: String? ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerLoad") { it.loadFromJs(url, headers, paused, title, artist, artworkUrl) }
    }

    AsyncFunction("nativePlayerSetRate") { rate: Double ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetRate") { it.setRateFromJs(rate) }
    }

    AsyncFunction("nativePlayerSetVolume") { volume: Double ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetVolume") { it.setVolumeFromJs(volume) }
    }

    AsyncFunction("nativePlayerSetResizeMode") { mode: String? ->
      Log.i("CrispyModule", "nativePlayerSetResizeMode mode=$mode")
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetResizeMode") { it.setResizeModeFromJs(mode) }
    }

    AsyncFunction("nativePlayerSetAudioTrack") { trackId: Int ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetAudioTrack") { it.setAudioTrackFromJs(trackId) }
    }

    AsyncFunction("nativePlayerSetSubtitleTrack") { trackId: Int ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleTrack") { it.setSubtitleTrackFromJs(trackId) }
    }

    AsyncFunction("nativePlayerSetSubtitleDelay") { delaySec: Double ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleDelay") { it.setSubtitleDelayFromJs(delaySec) }
    }

    // VLC/MPV compatibility stubs
    AsyncFunction("nativePlayerSetSubtitleSize") { size: Int ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleSize") { it.setSubtitleSizeFromJs(size) }
    }

    AsyncFunction("nativePlayerSetSubtitleColor") { color: String ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleColor") { it.setSubtitleColorFromJs(color) }
    }

    AsyncFunction("nativePlayerSetSubtitleBackgroundColor") { color: String, opacity: Float ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleBackgroundColor") { it.setSubtitleBackgroundColorFromJs(color, opacity) }
    }

    AsyncFunction("nativePlayerSetSubtitleBorderSize") { size: Int ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleBorderSize") { it.setSubtitleBorderSizeFromJs(size) }
    }

    AsyncFunction("nativePlayerSetSubtitleBorderColor") { color: String ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleBorderColor") { it.setSubtitleBorderColorFromJs(color) }
    }

    AsyncFunction("nativePlayerSetSubtitlePosition") { pos: Int ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitlePosition") { it.setSubtitlePositionFromJs(pos) }
    }

    AsyncFunction("nativePlayerSetSubtitleBold") { bold: Boolean ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleBold") { it.setSubtitleBoldFromJs(bold) }
    }

    AsyncFunction("nativePlayerSetSubtitleItalic") { italic: Boolean ->
      return@AsyncFunction withPlayerActivityUi("nativePlayerSetSubtitleItalic") { it.setSubtitleItalicFromJs(italic) }
    }

    // --- VIDEO PLAYER VIEW ---
    View(CrispyVlcVideoView::class) {
      Prop("source") { view: CrispyVlcVideoView, url: String? ->
        view.setSource(url)
      }

      Prop("headers") { view: CrispyVlcVideoView, headers: Map<String, String>? ->
        view.setHeaders(headers)
      }

      Prop("paused") { view: CrispyVlcVideoView, paused: Boolean ->
        view.setPaused(paused)
      }

      Prop("resizeMode") { view: CrispyVlcVideoView, mode: String? ->
        view.setResizeMode(mode)
      }
      
      Prop("playInBackground") { view: CrispyVlcVideoView, playInBackground: Boolean ->
        view.setPlayInBackground(playInBackground)
      }

      Events("onLoad", "onProgress", "onEnd", "onError", "onTracksChanged")

      AsyncFunction("seek") { view: CrispyVlcVideoView, positionSec: Double ->
        view.seek(positionSec)
      }

      AsyncFunction("enterPiP") { _: CrispyVlcVideoView ->
        val activity = appContext.currentActivity ?: return@AsyncFunction false
        return@AsyncFunction PipController.enterPiP(activity, null, null)
      }

      AsyncFunction("setAudioTrack") { view: CrispyVlcVideoView, trackId: Int ->
        view.setAudioTrack(trackId)
      }

      AsyncFunction("setSubtitleTrack") { view: CrispyVlcVideoView, trackId: Int ->
        view.setSubtitleTrack(trackId)
      }

      AsyncFunction("setSubtitleDelay") { view: CrispyVlcVideoView, delay: Double ->
        view.setSubtitleDelay(delay)
      }
      
      AsyncFunction("setMetadata") { view: CrispyVlcVideoView, title: String, artist: String, artworkUrl: String? ->
        view.setMetadata(title, artist, artworkUrl)
      }

      // Compat Stubs
      AsyncFunction("setSubtitleSize") { view: CrispyVlcVideoView, size: Int -> view.setSubtitleSize(size) }
      AsyncFunction("setSubtitleColor") { view: CrispyVlcVideoView, color: String -> view.setSubtitleColor(color) }
      AsyncFunction("setSubtitleBackgroundColor") { view: CrispyVlcVideoView, color: String, opacity: Float -> view.setSubtitleBackgroundColor(color, opacity) }
      AsyncFunction("setSubtitleBorderSize") { view: CrispyVlcVideoView, size: Int -> view.setSubtitleBorderSize(size) }
      AsyncFunction("setSubtitleBorderColor") { view: CrispyVlcVideoView, color: String -> view.setSubtitleBorderColor(color) }
      AsyncFunction("setSubtitlePosition") { view: CrispyVlcVideoView, pos: Int -> view.setSubtitlePosition(pos) }
      AsyncFunction("setSubtitleBold") { view: CrispyVlcVideoView, bold: Boolean -> view.setSubtitleBold(bold) }
      AsyncFunction("setSubtitleItalic") { view: CrispyVlcVideoView, italic: Boolean -> view.setSubtitleItalic(italic) }

      Prop("metadata") { view: CrispyVlcVideoView, metadata: Map<String, Any>? ->
        metadata?.let {
          val title = it["title"] as? String ?: ""
          val artist = (it["artist"] as? String) ?: (it["subtitle"] as? String) ?: ""
          val artworkUrl = it["artworkUrl"] as? String
          view.setMetadata(title, artist, artworkUrl)
        }
      }
    }
  }


  private fun ensureService(): TorrentService? {
      if (torrentService != null) return torrentService
      
      // Start and bind if not already bound
      if (!isBound) {
          val context = appContext.reactContext ?: return null
          Log.d("CrispyModule", "Starting TorrentService lazily...")
          val intent = Intent(context, TorrentService::class.java)
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
              context.startForegroundService(intent)
          } else {
              context.startService(intent)
          }
          context.bindService(intent, connection, Context.BIND_AUTO_CREATE)
      }
      
      return if (waitForService()) torrentService else null
  }

  private fun waitForService(): Boolean {
      if (torrentService != null) return true
      
      Log.d("CrispyModule", "Waiting for TorrentService...")
      try {
          // Wait up to 5 seconds for the service to bind
          val connected = serviceLatch.await(5, TimeUnit.SECONDS)
          if (connected) {
              Log.d("CrispyModule", "Service connected after wait")
          } else {
              Log.w("CrispyModule", "Service connection timed out")
          }
          return connected
      } catch (e: InterruptedException) {
          Log.e("CrispyModule", "Interrupted while waiting for service", e)
          return false
      }
  }

  private fun waitForStartupCleanup(timeoutMs: Long = 60_000L) {
      try {
          val completed = startupCleanupLatch.await(timeoutMs, TimeUnit.MILLISECONDS)
          if (!completed) {
              Log.w("CrispyModule", "Startup torrent cleanup timed out after ${timeoutMs}ms; continuing")
          }
      } catch (e: InterruptedException) {
          Log.e("CrispyModule", "Interrupted while waiting for startup cleanup", e)
      }
  }

  private fun withPlayerActivityUi(action: String, fn: (PlayerActivity) -> Unit): Boolean {
    val current = appContext.currentActivity
    val activity = (current as? PlayerActivity) ?: PlayerActivity.getActive()
    if (activity == null) {
      Log.w("CrispyModule", "$action: no PlayerActivity (current=${current?.javaClass?.simpleName})")
      return false
    }
    activity.runOnUiThread {
      try {
        fn(activity)
      } catch (t: Throwable) {
        Log.e("CrispyModule", "$action failed", t)
      }
    }
    return true
  }
}

File: modules/crispy-native-core/ios/CrispyNativeCoreModule.swift
import ExpoModulesCore
import AVKit
import AVFoundation
import MobileVLCKit

public class CrispyNativeCoreModule: Module {
    private var pipController: AVPictureInPictureController?
    
    public func definition() -> ModuleDefinition {
        Name("CrispyNativeCore")
        
        // MARK: - Torrent Methods (No-op on iOS)
        
        AsyncFunction("startStream") { (infoHash: String, fileIdx: Int, sessionId: String) -> String? in
            print("[CrispyNativeCore] Torrent streaming not supported on iOS")
            return nil
        }
        
        AsyncFunction("destroyStream") { (sessionId: String) -> Void in
            print("[CrispyNativeCore] destroyStream: Torrent not supported on iOS")
        }
        
        AsyncFunction("stopTorrent") { (infoHash: String) -> Void in
            print("[CrispyNativeCore] stopTorrent: Torrent not supported on iOS")
        }
        
        AsyncFunction("destroyTorrent") { (infoHash: String) -> Void in
            print("[CrispyNativeCore] destroyTorrent: Torrent not supported on iOS")
        }
        
        AsyncFunction("clearCache") { () -> Void in
            print("[CrispyNativeCore] clearCache: Torrent not supported on iOS")
        }
        
        AsyncFunction("handleSeek") { (infoHash: String, fileIdx: Int, position: Double) -> Void in
            print("[CrispyNativeCore] handleSeek: Torrent not supported on iOS")
        }
        
        // MARK: - Picture-in-Picture
        
        AsyncFunction("enterPiP") { (width: Double?, height: Double?) -> Void in
            guard AVPictureInPictureController.isPictureInPictureSupported() else {
                print("[CrispyNativeCore] PiP not supported on this device")
                return
            }
            
            DispatchQueue.main.async {
                self.enterPiPMode()
            }
        }
        
        AsyncFunction("setPiPConfig") { (enabled: Bool, isPlaying: Bool, width: Double?, height: Double?) -> Bool in
            return AVPictureInPictureController.isPictureInPictureSupported()
        }
        
        AsyncFunction("isInPiPMode") { () -> Bool in
            return pipController?.isPictureInPictureActive ?? false
        }
        
        // MARK: - Native Player Activity (Stub for iOS - handled differently)
        
        AsyncFunction("openPlayerActivity") { (
            sessionId: String,
            url: String,
            headers: [String: String]?,
            engine: String,
            paused: Bool,
            title: String,
            subtitle: String,
            artworkUrl: String?
        ) -> Bool in
            print("[CrispyNativeCore] openPlayerActivity: Use JS player on iOS")
            return false
        }
        
        AsyncFunction("closePlayerActivity") { () -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetPaused") { (paused: Bool) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSeek") { (positionSec: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerLoad") { (
            url: String?,
            headers: [String: String]?,
            paused: Bool,
            title: String,
            subtitle: String,
            artworkUrl: String?
        ) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetRate") { (rate: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetVolume") { (volume: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetResizeMode") { (mode: String) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetAudioTrack") { (trackId: Int) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleTrack") { (trackId: Int) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleDelay") { (delaySec: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleSize") { (size: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleColor") { (color: String) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleBackgroundColor") { (color: String, opacity: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleBorderSize") { (size: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleBorderColor") { (color: String) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitlePosition") { (pos: Double) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleBold") { (bold: Bool) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetSubtitleItalic") { (italic: Bool) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetDecoderMode") { (mode: String) -> Bool in
            return false
        }
        
        AsyncFunction("nativePlayerSetGpuMode") { (mode: String) -> Bool in
            return false
        }
        
        // MARK: - VLC View (primary iOS implementation)
        
        View(CrispyVLCVideoView.self) {
            Prop("source") { (view: CrispyVLCVideoView, source: String?) in
                view.setSource(source)
            }
            
            Prop("headers") { (view: CrispyVLCVideoView, headers: [String: String]?) in
                view.setHeaders(headers)
            }
            
            Prop("paused") { (view: CrispyVLCVideoView, paused: Bool) in
                view.setPaused(paused)
            }
            
            Prop("resizeMode") { (view: CrispyVLCVideoView, mode: String?) in
                view.setResizeMode(mode ?? "contain")
            }
            
            Prop("decoderMode") { (view: CrispyVLCVideoView, mode: String?) in
                view.setDecoderMode(mode ?? "auto")
            }
            
            Prop("gpuMode") { (view: CrispyVLCVideoView, mode: String?) in
                view.setGpuMode(mode ?? "gpu")
            }
            
            Prop("metadata") { (view: CrispyVLCVideoView, metadata: [String: String]?) in
                view.setMetadata(metadata)
            }
            
            Prop("playInBackground") { (view: CrispyVLCVideoView, enabled: Bool) in
                view.setPlayInBackground(enabled)
            }
            
            OnViewDidUpdateProps { (view: CrispyVLCVideoView) in
                view.applyProps()
            }
        }
    }
    
    private func enterPiPMode() {
        guard let window = UIApplication.shared.keyWindow,
              let rootViewController = window.rootViewController else {
            return
        }
        
        // Find any active AVPlayerLayer
        if let playerLayer = findPlayerLayer(in: rootViewController.view) {
            if pipController == nil {
                pipController = AVPictureInPictureController(playerLayer: playerLayer)
            }
            pipController?.startPictureInPicture()
        }
    }
    
    private func findPlayerLayer(in view: UIView) -> AVPlayerLayer? {
        for subview in view.subviews {
            if let playerLayer = subview.layer as? AVPlayerLayer {
                return playerLayer
            }
            if let found = findPlayerLayer(in: subview) {
                return found
            }
        }
        return nil
    }
}

// MARK: - VLC Video View

public class CrispyVLCVideoView: UIView {
    private var mediaPlayer: VLCMediaPlayer?
    private var videoView: UIView?
    private var displayLink: CADisplayLink?
    
    // Props
    private var source: String?
    private var headers: [String: String]?
    private var paused: Bool = false
    private var resizeMode: String = "contain"
    private var decoderMode: String = "auto"
    private var gpuMode: String = "gpu"
    private var metadata: [String: String]?
    private var playInBackground: Bool = false
    
    private var hasLoaded: Bool = false
    private var duration: Int = 0
    private var hasEmittedLoad: Bool = false
    
    // Callbacks for event emission
    var onLoad: (([String: Any]) -> Void)?
    var onProgress: (([String: Any]) -> Void)?
    var onEnd: (() -> Void)?
    var onError: ((String) -> Void)?
    var onBuffering: ((Bool) -> Void)?
    var onReadyForDisplay: (() -> Void)?
    var onTracksChanged: ((Any) -> Void)?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupView()
    }
    
    private func setupView() {
        backgroundColor = .black
        
        let video = UIView(frame: bounds)
        video.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        video.backgroundColor = .clear
        addSubview(video)
        self.videoView = video
        
        mediaPlayer = VLCMediaPlayer()
        mediaPlayer?.delegate = self
        mediaPlayer?.drawable = video
        
        // Setup display link for progress updates
        displayLink = CADisplayLink(target: self, selector: #selector(updateProgress))
        displayLink?.add(to: .main, forMode: .common)
    }
    
    public func setSource(_ source: String?) {
        self.source = source
        hasEmittedLoad = false
    }
    
    public func setHeaders(_ headers: [String: String]?) {
        self.headers = headers
    }
    
    public func setPaused(_ paused: Bool) {
        self.paused = paused
        guard hasLoaded else { return }
        
        if paused {
            mediaPlayer?.pause()
        } else {
            mediaPlayer?.play()
        }
    }
    
    public func setResizeMode(_ mode: String) {
        self.resizeMode = mode
    }
    
    public func setDecoderMode(_ mode: String) {
        self.decoderMode = mode
    }
    
    public func setGpuMode(_ mode: String) {
        self.gpuMode = mode
    }
    
    public func setMetadata(_ metadata: [String: String]?) {
        self.metadata = metadata
    }
    
    public func setPlayInBackground(_ enabled: Bool) {
        self.playInBackground = enabled
        configureAudioSession()
    }
    
    public func applyProps() {
        loadSource()
    }
    
    private func loadSource() {
        guard let source = source, let url = URL(string: source) else {
            return
        }
        
        let media = VLCMedia(url: url)
        
        // Add headers if provided
        if let headers = headers {
            for (key, value) in headers {
                media.addOptions(["http-\(key)=\(value)"])
            }
        }
        
        mediaPlayer?.media = media
        configureAudioSession()
        
        if !paused {
            mediaPlayer?.play()
        }
        
        hasLoaded = true
    }
    
    private func configureAudioSession() {
        let audioSession = AVAudioSession.sharedInstance()
        do {
            let category: AVAudioSession.Category = playInBackground ? .playback : .ambient
            try audioSession.setCategory(category, mode: .moviePlayback, options: [.allowAirPlay, .allowBluetooth])
            try audioSession.setActive(true)
        } catch {
            print("[CrispyVLCVideoView] Audio session error: \(error)")
        }
    }
    
    @objc private func updateProgress() {
        guard let player = mediaPlayer else { return }
        
        let currentTimeMs = Int(player.time.intValue)
        let durationMs = Int(player.media?.length.intValue ?? 0)
        
        // Emit onLoad once we have duration
        if durationMs > 0 && !hasEmittedLoad {
            self.duration = durationMs
            hasEmittedLoad = true
            onLoad?([
                "duration": durationMs / 1000,
                "width": 0,
                "height": 0
            ])
        }
        
        onProgress?([
            "position": currentTimeMs / 1000,
            "duration": durationMs / 1000
        ])
    }
    
    // MARK: - Public Methods
    
    public func seek(to positionSec: Double) {
        let targetTime = Int(positionSec * 1000)
        mediaPlayer?.time = VLCTime(int: Int32(targetTime))
    }
    
    public func setAudioTrack(_ trackId: Int) {
        let tracks = mediaPlayer?.audioTracksIndexes ?? []
        if trackId < tracks.count {
            mediaPlayer?.currentAudioTrackIndex = tracks[trackId]
        }
    }
    
    public func setSubtitleTrack(_ trackId: Int) {
        let subtitles = mediaPlayer?.videoSubTitlesIndexes ?? []
        if trackId < subtitles.count {
            mediaPlayer?.currentVideoSubTitleIndex = subtitles[trackId]
        }
    }
    
    public func addExternalSubtitle(url: String, title: String?, lang: String?) {
        // VLC iOS doesn't support adding external subtitles at runtime via public API
    }
    
    public func setSubtitleDelay(_ delay: Double) {
        let delayMs = Int(delay * 1000)
        mediaPlayer?.currentVideoSubTitleDelay = Int32(delayMs)
    }
    
    public func setSubtitleSize(_ size: Double) {}
    public func setSubtitleColor(_ color: String) {}
    public func setSubtitleBackgroundColor(_ color: String, opacity: Double) {}
    public func setSubtitleBorderSize(_ size: Double) {}
    public func setSubtitleBorderColor(_ color: String) {}
    public func setSubtitlePosition(_ pos: Double) {}
    public func setSubtitleBold(_ bold: Bool) {}
    public func setSubtitleItalic(_ italic: Bool) {}
    public func setDecoderMode(_ mode: String) {}
    public func setGpuMode(_ mode: String) {}
    
    public override func removeFromSuperview() {
        displayLink?.invalidate()
        displayLink = nil
        mediaPlayer?.stop()
        mediaPlayer?.delegate = nil
        mediaPlayer = nil
        super.removeFromSuperview()
    }
}

// MARK: - VLCMediaPlayerDelegate

extension CrispyVLCVideoView: VLCMediaPlayerDelegate {
    public func mediaPlayerStateChanged(_ aNotification: Notification) {
        guard let player = mediaPlayer else { return }
        
        switch player.state {
        case .buffering:
            onBuffering?(true)
        case .playing:
            onBuffering?(false)
            onReadyForDisplay?()
        case .ended:
            onEnd?()
        case .error:
            onError?("VLC playback error")
        default:
            break
        }
    }
    
    public func mediaPlayerTimeChanged(_ aNotification: Notification) {
        // Handled by display link for smoother updates
    }
}
